k2/csrc/array.h

```diff
@@ -454,17 +529,57 @@ class Array2 {
       with the "no-gaps" semantics if dim0_ <= 1. */
   bool IsContiguous() const { return elem_stride0_ == dim1_; }
 
+
+  // Only if this is a Array2<Any>: convert to a specific type, which must be
+  // identical to the actual type stored as dtype_.
+  // You call this as, e.g.
+  //  Array2<Any> a(...);
+  //  if (a.GetDtype() == kInt32Dtype) {
+  //     Array2<int32_t> &b = a.template Specialize<int32_t>();
+  //     ...
+  //  }
+  // Note: the '.template' above is only necessary if you are inside
+  // a template definition already.  It is an odd syntax.
+  template <typename U>
+  Array2<U> &Specialize() {
+    static_assert(std::is_same<T, Any>::value,
+                  "generic arrays not supported here");
+    K2_CHECK_EQ(dtype_, DtypeOf<U>::dtype);
+    return *reinterpret_cast<Array2<U>*>(this);
+  }
+  template <typename U>
+  const Array2<U> &Specialize() const {
+    static_assert(std::is_same<T, Any>::value,
+                  "generic arrays not supported here");
+    K2_CHECK_EQ(dtype_, DtypeOf<U>::dtype);
+    return *reinterpret_cast<const Array2<U>*>(this);
+  }
+
+  // Return a reference to this viewed as `Any` type (for when we
+  // want a generic array without type informatiopn)
+  // Reverse this with Specialize().
+  Array2<Any> &Generic() {
+    return *reinterpret_cast<Array2<Any>*>(this);
+  }
+  Array2<Any> &Generic() const {
```

Comment:
Since its return type is not a const lvalue reference, I am afraid
it will result in a compile error if you use this function.

The compiler does not complain at present because you are not using it.